cmake/Modules/GetVersionFromGit.cmake

```diff
@@ -0,0 +1,109 @@
+find_package(Git REQUIRED)
+
+# Attempt to retrieve version from Git
+if(EXISTS "${CMAKE_SOURCE_DIR}/.git" AND GIT_FOUND)
+    message(STATUS "Using git describe for versioning")
+
+    # Extract the version string
+    execute_process(
+        COMMAND git describe --tags --dirty
+        WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+        OUTPUT_VARIABLE VERSION_STRING
+        OUTPUT_STRIP_TRAILING_WHITESPACE
+    )
+
+    # Extract the commit hash
+    execute_process(
+        COMMAND git rev-parse HEAD
+        WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+        OUTPUT_VARIABLE COMMIT_HASH
+        OUTPUT_STRIP_TRAILING_WHITESPACE
+    )
+else()
+    message(STATUS "Using archival file for versioning: ${ARCHIVAL_FILE}")
+    if(EXISTS "${ARCHIVAL_FILE}")
+        file(READ "${ARCHIVAL_FILE}" ARCHIVAL_CONTENT)
+
+        # Extract the describe-name line
+        string(REGEX MATCH "${DESCRIBE_NAME_KEY}([^\\n]+)" VERSION_STRING "${ARCHIVAL_CONTENT}")
+        if(VERSION_STRING MATCHES "${DESCRIBE_NAME_KEY}(.*)")
+            set(VERSION_STRING "${CMAKE_MATCH_1}")
+        else()
+            message(FATAL_ERROR "Could not extract version from ${ARCHIVAL_FILE}")
+        endif()
+
+        # Extract the commit hash
+        string(REGEX MATCH "${COMMIT_HASH_KEY}([a-f0-9]+)" COMMIT_HASH "${ARCHIVAL_CONTENT}")
+        if(COMMIT_HASH MATCHES "${COMMIT_HASH_KEY}([a-f0-9]+)")
+            set(COMMIT_HASH "${CMAKE_MATCH_1}")
+        else()
+            message(FATAL_ERROR "Could not extract commit hash from ${ARCHIVAL_FILE}")
+        endif()
+    else()
+        message(FATAL_ERROR "Neither git describe nor ${ARCHIVAL_FILE} is available for versioning.")
+    endif()
+endif()
+
+# Ensure version string format
+if(VERSION_STRING MATCHES "${VERSION_REGEX_WITH_PREFIX}")
+    set(VERSION_NO_SUFFIX "${CMAKE_MATCH_1}")
+else()
+    message(FATAL_ERROR "Invalid version format: Missing base version in ${VERSION_STRING}")
+endif()
+
+# Check for development state
+if(VERSION_STRING MATCHES "-([0-9]+)-g([0-9a-f]+)")
+    set(DEV_STATE "true")
+    set(COMMIT_COUNT "${CMAKE_MATCH_1}")
+    string(REGEX REPLACE "-([0-9]+)-g([0-9a-f]+)" "" VERSION_WITHOUT_META "${VERSION_STRING}")
+else()
+    set(DEV_STATE "false")
+    set(VERSION_WITHOUT_META "${VERSION_STRING}")
+endif()
+
+# Split and set version components
+string(REPLACE "." ";" VERSION_LIST "${VERSION_NO_SUFFIX}")
+list(GET VERSION_LIST 0 VERSION_MAJOR)
+list(GET VERSION_LIST 1 VERSION_MINOR)
+list(GET VERSION_LIST 2 VERSION_PATCH)
+
+# Increment patch number for dev versions
+if(DEV_STATE)
+    math(EXPR VERSION_PATCH "${VERSION_PATCH} + 1")
+endif()
+
+# Export variables
+set(OPENMC_VERSION_MAJOR "${VERSION_MAJOR}")
+set(OPENMC_VERSION_MINOR "${VERSION_MINOR}")
+set(OPENMC_VERSION_PATCH "${VERSION_PATCH}")
+set(OPENMC_VERSION "${VERSION_MAJOR}.${VERSION_MINOR}.${VERSION_PATCH}")
+set(OPENMC_COMMIT_HASH "${COMMIT_HASH}")
+set(OPENMC_DEV_STATE "${DEV_STATE}")
+set(OPENMC_COMMIT_COUNT "${COMMIT_COUNT}")
```

Comment:
FWIW, I thought the `PROJECT_` variable names made sense if one imagined this CMake file being easily ported to other projects....

Still possible with a quick search/replace, but nice to have it be otherwise generic (as someone who wants to quickly reuse this elsewhere)

Wouldn't hold up this merge over it, though...

Reply:
My understanding is that `PROJECT_*` variables should not be set manually; normally that is the job of the call to `project(...)`.

Reply:
I was just arguing (weakly) for something more generic, but not specifically `PROJECT_` (since I didn't know about that reservation on its use)

Reply:
I personally prefer keeping it simple with fewer lines of code when possible. If another project wants to use this, it would be very easy to adapt it.

Reply:
👍 - thus my thought about not holding up a merge....